src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -266,5 +273,33 @@ protected virtual string GenerateUrl(string protocol, string host, VirtualPathDa
                 builder.Clear();
             }
         }
+
+        private string FastGetUrl(
+            string protocol, 
+            string host, 
+            VirtualPathData pathData, 
+            string fragment)
+        {
+            if (string.IsNullOrEmpty(protocol) 
+                && string.IsNullOrEmpty(host) 
+                && string.IsNullOrEmpty(fragment))
```

Comment:
These checks are copied from the original code path https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs#L239